### ISSUE TITLE
training with my own dataset

### ISSUE THREAD

**Issue by messithanh2k** [davidnvq/grit#26]
Thank authors. Your contributions is wonderful.
Can you help me how to train with my own dataset. Thank you!

**Comment by davidnvq** [comment]
Thanks for your interest. I may help if you give me more detail about your own dataset, for example the annotation format?

**Comment by messithanh2k** [comment]
thanks for reply. Are you Vietnamese. I'm from Vietnam.
I use dataset in vietnamese only include images folder and json caption file. Thank you

**Comment by davidnvq** [comment]
Hi there, I'm a Vietnamese, too.
Working on Vietnamese language requires extra effort. I'm willing to help or collaborate. Shoot me an email (to quang@vision.is.tohoku.ac.jp) for more detail about your task.

**Comment by victorup** [comment]
Hi,
I also want to train on my own dataset. The format of my dataset looks like that:
```
1. image_path1 caption1
2. image_path2 caption2
3. ...
```
What should I do to train the model on these image-text pairs?
Thanks!

**Comment by davidnvq** [comment]
Hi there, 

Sorry for the late reply.

If your dataset consists of English-based captions, I strongly recommend to finetune the given checkpoint on COCO dataset (check the README file in order to download). Therefore, you can make use of `vocab.json` generated from the COCO dataset, otherwise, you need to create your own `vocab.json`.

Based on your given formats, I have made a workable version as follows.

## Config
Just duplicate the `coco_config.yaml` in `configs/caption` as `yourdataset_config.yaml`. For some parameters such as finetuning COCO-based checkpoint and freezing backbone, etc, I recommend the following:
```
exp:
  ngpus_per_node: 2 # <- This is based on N-GPUs you have
  world_size: 2 # <- If you have 1 machine/cluster, then it is equal to ngpus_per_node
  checkpoint: 'path/to/grit_checkpoint_4ds.pth' # Check README to download it.
...

optimizer:
  num_epochs: 20 # <- Add this to simplify the training strategy
  freeze_detector: True #  <- Speed up your training
  freeze_backbone: True # <- Speed up your training 

```

## Datasets
Currently, I assume that one image has ONE caption (it means that `image_pathN` is different from `image_pathM`). Otherwise, you need make some further effort so that `gold_captions` in `YourCollator` would include all captions of ONE image to compute scores of the metrics accurately.

I suggest you to create a file  `yourdataset.py`  (for example: `./datasets/caption/yourdataset.py`) as follows:
```python
import os
import json
import torch
from PIL import Image
from engine.utils import nested_tensor_from_tensor_list
from torch.nn.utils.rnn import pad_sequence


class YourDataset:

    def __init__(self, ann_path, vocab=None, transform=None):
        self.data = json.load(open(ann_path))
        self.vocab = vocab
        self.transform = transform

    def __getitem__(self, index: int):
        item = self.data[index]

        # read image
        image = Image.open(item[0])
        image = self.transform(image)

        caption = item[1]
        return {'image': image, 'caption': item[1]}

    def __len__(self):
        return len(self.data)


class YourCollator:

    def __init__(self, text_field, device='cpu', max_len=54, pad_idx=1, bos_idx=2, eos_idx=3):
        self.device = device
        # from coco
        self.max_len = max_len
        self.pad_idx = pad_idx
        self.bos_idx = bos_idx
        self.eos_idx = eos_idx

        self.text_field = text_field  # from coco

    def truncate_pad(self, captions):

        raw_captions = [[self.text_field.vocab.stoi[w] for w in self.text_field.preprocess(seq)] for seq in captions]

        # truncate
        captions = [s[:self.max_len] for s in raw_captions]

        # padding
        max_len = max([len(c) for c in captions])
        padded_captions = []
        for c in captions:
            seq = [self.bos_idx] + c + [self.eos_idx] + [self.pad_idx] * (max_len - len(c))
            padded_captions.append(seq)

        padded_captions = [torch.Tensor(seq).long() for seq in padded_captions]
        padded_captions = pad_sequence(padded_captions, batch_first=True).to(self.device)

        return raw_captions, padded_captions

    def __call__(self, batch):
        images = nested_tensor_from_tensor_list([item['image'] for item in batch]).to(self.device)
        captions = [item['caption'] for item in batch]
        raw_captions, padded_captions = self.truncate_pad(captions)

        return {
            "samples": images,  # images.mask: 1 for MASKED pixels; 0 for NOT masked tokens
            "gold_captions": [[caption] for caption in raw_captions],  # for computing metrics' scores
            "captions": padded_captions,
        }
```
## Training and Evaluation
Roughly, I will create a simple code for training with Cross-Entropy Loss only. If you want to further finetune using reinforcement learning, please refer to `train_sc`  in `engine/caption_engine.py` and adapt the code wisely:

```python
import os
import hydra
import random
import numpy as np
from omegaconf import DictConfig

from datasets.caption.field import TextField
from models.caption import Transformer, GridFeatureNetwork, CaptionGenerator
from models.caption.detector import build_detector
from models.common.attention import MemoryAttention
from utils.cap_scheduler import CosineLRScheduler

import torch
import torch.distributed as dist
import torch.multiprocessing as mp
from torch.nn.parallel import DistributedDataParallel as DDP
from torch.utils.tensorboard import SummaryWriter
from engine.sherlock_engine import *
from torch.utils.data import DataLoader, DistributedSampler, BatchSampler


def main(gpu, config):
    # dist init
    torch.backends.cudnn.enabled = False
    rank = config.exp.rank * config.exp.ngpus_per_node + gpu
    dist.init_process_group('nccl', 'env://', rank=rank, world_size=config.exp.world_size)

    torch.manual_seed(config.exp.seed)
    np.random.seed(config.exp.seed)
    random.seed(config.exp.seed)

    device = torch.device(f"cuda:{gpu}")
    torch.cuda.set_device(gpu)

    # extract features
    detector = build_detector(config).to(device)

    if os.path.exists(config.model.detector.checkpoint):
        checkpoint = torch.load(config.model.detector.checkpoint, map_location='cpu')
        missing, unexpected = detector.load_state_dict(checkpoint['model'], strict=False)
        print("det missing:", len(missing), missing)
        print("det unexpected:", len(unexpected), unexpected)

    grit_net = GridFeatureNetwork(
        pad_idx=config.model.pad_idx,
        d_in=config.model.grid_feat_dim,
        dropout=config.model.dropout,
        attn_dropout=config.model.attn_dropout,
        attention_module=MemoryAttention,
        **config.model.grit_net,
    )
    cap_generator = CaptionGenerator(
        vocab_size=config.model.vocab_size,
        max_len=config.model.max_len,
        pad_idx=config.model.pad_idx,
        dropout=config.model.dropout,
        attn_dropout=config.model.attn_dropout,
        cfg=config.model.cap_generator,
        **config.model.cap_generator,
    )
    model = Transformer(
        grit_net,
        cap_generator,
        detector=detector,  # .module,
        use_gri_feat=config.model.use_gri_feat,
        use_reg_feat=config.model.use_reg_feat,
        config=config,
    )

    model = model.to(device)

    if os.path.exists(config.exp.checkpoint):
        checkpoint = torch.load(config.exp.checkpoint, map_location='cpu')
        missing, unexpected = model.load_state_dict(checkpoint['state_dict'], strict=False)
        print("model missing:", len(missing))
        print("model unexpected:", len(unexpected))
        if 'backbone' in checkpoint:
            model.detector.backbone.load_state_dict(checkpoint['backbone'], strict=False)

    if getattr(config.optimizer, 'freeze_backbone', False):
        for n, p in model.named_parameters():
            if 'backbone' in n:
                p.requires_grad = False

    if getattr(config.optimizer, 'freeze_detector', False):
        for n, p in model.named_parameters():
            if 'detector' in n:
                p.requires_grad = False

    model = DDP(model, device_ids=[gpu], find_unused_parameters=True, broadcast_buffers=False)
    optimizers = build_optimizers(model, config, mode='xe')

    # tensorboard:
    writer = SummaryWriter(log_dir='tensorboard')

    # dataset
    from datasets.caption.yourdataset import YourDataset, YourCollator
    from datasets.caption.transforms import get_transform

    text_field = TextField(vocab_path=config.dataset.vocab_path)  # path to COCO vocab.json
    transforms = get_transform(config.dataset.transform_cfg)

    train_dataset = YourDataset(ann_path=train_ann_path, vocab=text_field.vocab, transform=transforms['train'])
    valid_dataset = YourDataset(ann_path=valid_ann_path, vocab=text_field.vocab, transform=transforms['valid'])

    collator = YourCollator(text_field=text_field, device=device)

    train_sampler = DistributedSampler(train_dataset, shuffle=True)
    batch_sampler = BatchSampler(train_sampler, batch_size=config.optimizer.batch_size, drop_last=True)

    dataloaders = {
        'train':
            DataLoader(train_dataset,
                       batch_sampler=batch_sampler,
                       num_workers=config.optimizer.num_workers,
                       collate_fn=collator),
        'valid':
            DataLoader(valid_dataset,
                       batch_size=16,
                       shuffle=False,
                       num_workers=config.optimizer.num_workers,
                       collate_fn=collator)
    }

    phase = 'ft_xe'
    scheduler = CosineLRScheduler(
        optimizers['model'],
        num_epochs=config.optimizer.num_epochs,
        num_its_per_epoch=len(dataloaders['train']),
        init_lr=config.optimizer.xe_lr,
        min_lr=config.optimizer.min_lr,
        warmup_init_lr=config.optimizer.warmup_init_lr,
    )

    for epoch in range(config.optimizer.num_epochs):
        print(f"Train: epoch={epoch}, phase={phase}")

        model.train()
        loss_fn = NLLLoss(ignore_index=text_field.vocab.stoi['<pad>'])

        if scheduler is not None:
            scheduler.step()

        running_loss = .0
        with tqdm(desc=f'Epoch {epoch} - train', unit='it', total=len(dataloaders['train'])) as pbar:
            for it, batch in enumerate(dataloaders['train']):
                out = model(batch['samples'], batch['captions'])
                optimizers['model'].zero_grad()
                optimizers['backbone'].zero_grad()

                captions_gt = batch['captions'][:, 1:].contiguous()
                out = out[:, :-1].contiguous()
                loss = loss_fn(out.view(-1, len(text_field.vocab)), captions_gt.view(-1))
                loss.backward()

                optimizers['model'].step()
                optimizers['backbone'].step()

                # todo: Not DDP
                loss = gather_result(loss)
                running_loss += loss.item()

                pbar.set_postfix(loss=running_loss / (it + 1))
                pbar.update()

                if scheduler is not None:
                    lr = scheduler.step()

                if rank == 0:
                    writer.add_scalar(
                        'backbone_lr',
                        optimizers['backbone'].param_groups[0]['lr'],
                        epoch * len(dataloaders['train']) + it,
                    )
                    writer.add_scalar(
                        'model_lr',
                        optimizers['model'].param_groups[0]['lr'],
                        epoch * len(dataloaders['train']) + it,
                    )
                    lr = optimizers['model'].param_groups[0]['lr']

        train_res = {
            'loss': running_loss / len(dataloaders['train']),
            'reward': 0,
            'reward_baseline': 0,
            'val_loss': 0.0,
        }

        if rank == 0:
            model.eval()
            gen, gts = {}, {}

            counter = 0
            times = []
            split = 'valid'
            dataloader = dataloaders[split]
            with tqdm(desc=f'Epoch {epoch} - evaluation on {split}', unit='it', total=len(dataloader)) as pbar:
                results = []
                for it, batch in enumerate(iter(dataloader)):
                    counter += 1
                    start_it = time.time()
                    with torch.no_grad():
                        out, _ = model(
                            batch['samples'],
                            seq=None,
                            use_beam_search=True,
                            max_len=config.model.beam_len,
                            eos_idx=config.model.eos_idx,
                            beam_size=config.model.beam_size,
                            out_size=1,
                            return_probs=False,
                        )
                    torch.cuda.synchronize()
                    end_it = time.time()
                    times.append(end_it - start_it)

                    if 'samples' in batch and not isinstance(batch['samples'], dict):
                        bs = batch['samples'].tensors.shape[0]
                    else:
                        bs = batch['samples']['reg_feat'].shape[0]
                    if it % 100 == 0:
                        print(
                            f"Number of iterations: {counter}, batch_size={bs}, Total time per 1 batch: {sum(times)/counter:0.5f}s"
                        )

                    caps_gen = text_field.decode(out, join_words=False)
                    # todo: turn batch['captions'] into list of (list of captions / image)
                    for i, (gts_i, gen_i) in enumerate(zip(batch['gold_captions'], caps_gen)):
                        gen_i = ' '.join([k for k, g in itertools.groupby(gen_i)])
                        gen[f'{it}_{i}'] = [gen_i]
                        gts[f'{it}_{i}'] = gts_i
                        res = {'image_id': batch['image_id'][i], 'caption': gen_i}
                        results.append(res)
                    pbar.update()

            avg_time = sum(times) / counter
            print(f"Epoch: {epoch} iters: {counter}\nTotal time per 1 batch: {avg_time:0.5f}s")
            gts = metrics.PTBTokenizer.tokenize(gts)
            gen = metrics.PTBTokenizer.tokenize(gen)
            scores, _ = metrics.compute_scores(gts, gen)
            print(f'Epoch {epoch}: {split} scores: ' + str(scores) + '\n')

            with open('result.txt', 'a') as f:
                f.write(f'Epoch {epoch}: {split} scores: ' + str(scores) + '\n')
            log_epoch(config, writer, epoch, train_res, split=split, scores=scores, which='ft_xe')

            torch.save(
                {
                    "state_dict":
                        model.module.state_dict() if getattr(model, 'module', None) is not None else model.state_dict(),
                    "optim_model":
                        optimizers['model'].state_dict(),
                    "optim_backbone":
                        optimizers['backbone'].state_dict(),
                    "epoch":
                        epoch,
                    "exp_name":
                        "" if config is None else config.exp.name,
                    "scheduler": [] if scheduler is None else scheduler.state_dict(),
                }, f'checkpoint_epoch_{epoch}.pth')
            torch.distributed.barrier()


@hydra.main(config_path="configs/caption", config_name="yourdataset_config") # <- it is to distinguish your config
def run_main(config: DictConfig) -> None:
    mp.spawn(main, nprocs=config.exp.ngpus_per_node, args=(config,))


if __name__ == "__main__":
    run_main()

```

**Comment by victorup** [comment]
@davidnvq Thank you so much! It is very useful to me.
